{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8c049e90-cd39-419b-8bc5-46717a73a8b6",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Installing Hail and the Google Cloud Storage filesystem connector on a GCP Vertex AI Workbench instance"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4111929d-1f5c-4e72-a4e6-11eb6c8a77e1",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Install the last version of Hail"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "14826859-1f80-4752-bbf3-cd7a3df753a9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: hail in /opt/conda/lib/python3.10/site-packages (0.2.128)\n",
      "Requirement already satisfied: aiodns<3,>=2.0.0 in /opt/conda/lib/python3.10/site-packages (from hail) (2.0.0)\n",
      "Requirement already satisfied: aiohttp<4,>=3.9 in /opt/conda/lib/python3.10/site-packages (from hail) (3.9.3)\n",
      "Requirement already satisfied: azure-identity<2,>=1.6.0 in /opt/conda/lib/python3.10/site-packages (from hail) (1.15.0)\n",
      "Requirement already satisfied: azure-mgmt-storage==20.1.0 in /opt/conda/lib/python3.10/site-packages (from hail) (20.1.0)\n",
      "Requirement already satisfied: azure-storage-blob<13,>=12.11.0 in /opt/conda/lib/python3.10/site-packages (from hail) (12.19.1)\n",
      "Requirement already satisfied: boto3<2.0,>=1.17 in /opt/conda/lib/python3.10/site-packages (from hail) (1.34.74)\n",
      "Requirement already satisfied: botocore<2.0,>=1.20 in /opt/conda/lib/python3.10/site-packages (from hail) (1.34.74)\n",
      "Requirement already satisfied: dill<0.4,>=0.3.6 in /opt/conda/lib/python3.10/site-packages (from hail) (0.3.8)\n",
      "Requirement already satisfied: frozenlist<2,>=1.3.1 in /opt/conda/lib/python3.10/site-packages (from hail) (1.4.1)\n",
      "Requirement already satisfied: google-auth<3,>=2.14.1 in /opt/conda/lib/python3.10/site-packages (from hail) (2.28.2)\n",
      "Requirement already satisfied: google-auth-oauthlib<1,>=0.5.2 in /opt/conda/lib/python3.10/site-packages (from hail) (0.8.0)\n",
      "Requirement already satisfied: humanize<2,>=1.0.0 in /opt/conda/lib/python3.10/site-packages (from hail) (1.1.0)\n",
      "Requirement already satisfied: janus<1.1,>=0.6 in /opt/conda/lib/python3.10/site-packages (from hail) (1.0.0)\n",
      "Requirement already satisfied: nest-asyncio<2,>=1.5.8 in /opt/conda/lib/python3.10/site-packages (from hail) (1.6.0)\n",
      "Requirement already satisfied: orjson<3.9.12,>=3.6.4 in /opt/conda/lib/python3.10/site-packages (from hail) (3.9.11)\n",
      "Requirement already satisfied: protobuf==3.20.2 in /opt/conda/lib/python3.10/site-packages (from hail) (3.20.2)\n",
      "Requirement already satisfied: rich<13,>=12.6.0 in /opt/conda/lib/python3.10/site-packages (from hail) (12.6.0)\n",
      "Requirement already satisfied: typer<1,>=0.9.0 in /opt/conda/lib/python3.10/site-packages (from hail) (0.9.0)\n",
      "Requirement already satisfied: python-json-logger<3,>=2.0.2 in /opt/conda/lib/python3.10/site-packages (from hail) (2.0.7)\n",
      "Requirement already satisfied: pyyaml<7.0,>=6.0 in /opt/conda/lib/python3.10/site-packages (from hail) (6.0.1)\n",
      "Requirement already satisfied: sortedcontainers<3,>=2.4.0 in /opt/conda/lib/python3.10/site-packages (from hail) (2.4.0)\n",
      "Requirement already satisfied: tabulate<1,>=0.8.9 in /opt/conda/lib/python3.10/site-packages (from hail) (0.9.0)\n",
      "Requirement already satisfied: jproperties<3,>=2.1.1 in /opt/conda/lib/python3.10/site-packages (from hail) (2.1.1)\n",
      "Requirement already satisfied: avro<1.12,>=1.10 in /opt/conda/lib/python3.10/site-packages (from hail) (1.11.3)\n",
      "Requirement already satisfied: bokeh<4,>=3 in /opt/conda/lib/python3.10/site-packages (from hail) (3.4.0)\n",
      "Requirement already satisfied: decorator<5 in /opt/conda/lib/python3.10/site-packages (from hail) (4.4.2)\n",
      "Requirement already satisfied: Deprecated<1.3,>=1.2.10 in /opt/conda/lib/python3.10/site-packages (from hail) (1.2.14)\n",
      "Requirement already satisfied: numpy<2 in /opt/conda/lib/python3.10/site-packages (from hail) (1.26.4)\n",
      "Requirement already satisfied: pandas<3,>=2 in /opt/conda/lib/python3.10/site-packages (from hail) (2.2.1)\n",
      "Requirement already satisfied: parsimonious<1 in /opt/conda/lib/python3.10/site-packages (from hail) (0.10.0)\n",
      "Requirement already satisfied: plotly<6,>=5.18.0 in /opt/conda/lib/python3.10/site-packages (from hail) (5.20.0)\n",
      "Requirement already satisfied: pyspark<3.4,>=3.3.2 in /opt/conda/lib/python3.10/site-packages (from hail) (3.3.4)\n",
      "Requirement already satisfied: requests<3,>=2.31.0 in /opt/conda/lib/python3.10/site-packages (from hail) (2.31.0)\n",
      "Requirement already satisfied: scipy<1.12,>1.2 in /opt/conda/lib/python3.10/site-packages (from hail) (1.11.4)\n",
      "Requirement already satisfied: uvloop<1,>=0.19.0 in /opt/conda/lib/python3.10/site-packages (from hail) (0.19.0)\n",
      "Requirement already satisfied: msrest>=0.6.21 in /opt/conda/lib/python3.10/site-packages (from azure-mgmt-storage==20.1.0->hail) (0.7.1)\n",
      "Requirement already satisfied: azure-common~=1.1 in /opt/conda/lib/python3.10/site-packages (from azure-mgmt-storage==20.1.0->hail) (1.1.28)\n",
      "Requirement already satisfied: azure-mgmt-core<2.0.0,>=1.3.1 in /opt/conda/lib/python3.10/site-packages (from azure-mgmt-storage==20.1.0->hail) (1.4.0)\n",
      "Requirement already satisfied: pycares>=3.0.0 in /opt/conda/lib/python3.10/site-packages (from aiodns<3,>=2.0.0->hail) (4.4.0)\n",
      "Requirement already satisfied: aiosignal>=1.1.2 in /opt/conda/lib/python3.10/site-packages (from aiohttp<4,>=3.9->hail) (1.3.1)\n",
      "Requirement already satisfied: attrs>=17.3.0 in /opt/conda/lib/python3.10/site-packages (from aiohttp<4,>=3.9->hail) (23.2.0)\n",
      "Requirement already satisfied: multidict<7.0,>=4.5 in /opt/conda/lib/python3.10/site-packages (from aiohttp<4,>=3.9->hail) (6.0.5)\n",
      "Requirement already satisfied: yarl<2.0,>=1.0 in /opt/conda/lib/python3.10/site-packages (from aiohttp<4,>=3.9->hail) (1.9.4)\n",
      "Requirement already satisfied: async-timeout<5.0,>=4.0 in /opt/conda/lib/python3.10/site-packages (from aiohttp<4,>=3.9->hail) (4.0.3)\n",
      "Requirement already satisfied: azure-core<2.0.0,>=1.23.0 in /opt/conda/lib/python3.10/site-packages (from azure-identity<2,>=1.6.0->hail) (1.30.1)\n",
      "Requirement already satisfied: cryptography>=2.5 in /opt/conda/lib/python3.10/site-packages (from azure-identity<2,>=1.6.0->hail) (42.0.5)\n",
      "Requirement already satisfied: msal<2.0.0,>=1.24.0 in /opt/conda/lib/python3.10/site-packages (from azure-identity<2,>=1.6.0->hail) (1.28.0)\n",
      "Requirement already satisfied: msal-extensions<2.0.0,>=0.3.0 in /opt/conda/lib/python3.10/site-packages (from azure-identity<2,>=1.6.0->hail) (1.1.0)\n",
      "Requirement already satisfied: typing-extensions>=4.3.0 in /opt/conda/lib/python3.10/site-packages (from azure-storage-blob<13,>=12.11.0->hail) (4.10.0)\n",
      "Requirement already satisfied: isodate>=0.6.1 in /opt/conda/lib/python3.10/site-packages (from azure-storage-blob<13,>=12.11.0->hail) (0.6.1)\n",
      "Requirement already satisfied: Jinja2>=2.9 in /opt/conda/lib/python3.10/site-packages (from bokeh<4,>=3->hail) (3.1.3)\n",
      "Requirement already satisfied: contourpy>=1.2 in /opt/conda/lib/python3.10/site-packages (from bokeh<4,>=3->hail) (1.2.0)\n",
      "Requirement already satisfied: packaging>=16.8 in /opt/conda/lib/python3.10/site-packages (from bokeh<4,>=3->hail) (24.0)\n",
      "Requirement already satisfied: pillow>=7.1.0 in /opt/conda/lib/python3.10/site-packages (from bokeh<4,>=3->hail) (10.2.0)\n",
      "Requirement already satisfied: tornado>=6.2 in /opt/conda/lib/python3.10/site-packages (from bokeh<4,>=3->hail) (6.4)\n",
      "Requirement already satisfied: xyzservices>=2021.09.1 in /opt/conda/lib/python3.10/site-packages (from bokeh<4,>=3->hail) (2023.10.1)\n",
      "Requirement already satisfied: jmespath<2.0.0,>=0.7.1 in /opt/conda/lib/python3.10/site-packages (from boto3<2.0,>=1.17->hail) (1.0.1)\n",
      "Requirement already satisfied: s3transfer<0.11.0,>=0.10.0 in /opt/conda/lib/python3.10/site-packages (from boto3<2.0,>=1.17->hail) (0.10.1)\n",
      "Requirement already satisfied: python-dateutil<3.0.0,>=2.1 in /opt/conda/lib/python3.10/site-packages (from botocore<2.0,>=1.20->hail) (2.9.0)\n",
      "Requirement already satisfied: urllib3!=2.2.0,<3,>=1.25.4 in /opt/conda/lib/python3.10/site-packages (from botocore<2.0,>=1.20->hail) (1.26.18)\n",
      "Requirement already satisfied: wrapt<2,>=1.10 in /opt/conda/lib/python3.10/site-packages (from Deprecated<1.3,>=1.2.10->hail) (1.16.0)\n",
      "Requirement already satisfied: cachetools<6.0,>=2.0.0 in /opt/conda/lib/python3.10/site-packages (from google-auth<3,>=2.14.1->hail) (5.3.3)\n",
      "Requirement already satisfied: pyasn1-modules>=0.2.1 in /opt/conda/lib/python3.10/site-packages (from google-auth<3,>=2.14.1->hail) (0.3.0)\n",
      "Requirement already satisfied: rsa<5,>=3.1.4 in /opt/conda/lib/python3.10/site-packages (from google-auth<3,>=2.14.1->hail) (4.9)\n",
      "Requirement already satisfied: requests-oauthlib>=0.7.0 in /opt/conda/lib/python3.10/site-packages (from google-auth-oauthlib<1,>=0.5.2->hail) (1.4.0)\n",
      "Requirement already satisfied: six~=1.13 in /opt/conda/lib/python3.10/site-packages (from jproperties<3,>=2.1.1->hail) (1.16.0)\n",
      "Requirement already satisfied: pytz>=2020.1 in /opt/conda/lib/python3.10/site-packages (from pandas<3,>=2->hail) (2024.1)\n",
      "Requirement already satisfied: tzdata>=2022.7 in /opt/conda/lib/python3.10/site-packages (from pandas<3,>=2->hail) (2024.1)\n",
      "Requirement already satisfied: regex>=2022.3.15 in /opt/conda/lib/python3.10/site-packages (from parsimonious<1->hail) (2023.12.25)\n",
      "Requirement already satisfied: tenacity>=6.2.0 in /opt/conda/lib/python3.10/site-packages (from plotly<6,>=5.18.0->hail) (8.2.3)\n",
      "Requirement already satisfied: py4j==0.10.9.5 in /opt/conda/lib/python3.10/site-packages (from pyspark<3.4,>=3.3.2->hail) (0.10.9.5)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /opt/conda/lib/python3.10/site-packages (from requests<3,>=2.31.0->hail) (3.3.2)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /opt/conda/lib/python3.10/site-packages (from requests<3,>=2.31.0->hail) (3.6)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /opt/conda/lib/python3.10/site-packages (from requests<3,>=2.31.0->hail) (2024.2.2)\n",
      "Requirement already satisfied: commonmark<0.10.0,>=0.9.0 in /opt/conda/lib/python3.10/site-packages (from rich<13,>=12.6.0->hail) (0.9.1)\n",
      "Requirement already satisfied: pygments<3.0.0,>=2.6.0 in /opt/conda/lib/python3.10/site-packages (from rich<13,>=12.6.0->hail) (2.17.2)\n",
      "Requirement already satisfied: click<9.0.0,>=7.1.1 in /opt/conda/lib/python3.10/site-packages (from typer<1,>=0.9.0->hail) (8.1.7)\n",
      "Requirement already satisfied: cffi>=1.12 in /opt/conda/lib/python3.10/site-packages (from cryptography>=2.5->azure-identity<2,>=1.6.0->hail) (1.16.0)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in /opt/conda/lib/python3.10/site-packages (from Jinja2>=2.9->bokeh<4,>=3->hail) (2.1.5)\n",
      "Requirement already satisfied: PyJWT<3,>=1.0.0 in /opt/conda/lib/python3.10/site-packages (from PyJWT[crypto]<3,>=1.0.0->msal<2.0.0,>=1.24.0->azure-identity<2,>=1.6.0->hail) (2.8.0)\n",
      "Requirement already satisfied: portalocker<3,>=1.0 in /opt/conda/lib/python3.10/site-packages (from msal-extensions<2.0.0,>=0.3.0->azure-identity<2,>=1.6.0->hail) (2.8.2)\n",
      "Requirement already satisfied: pyasn1<0.6.0,>=0.4.6 in /opt/conda/lib/python3.10/site-packages (from pyasn1-modules>=0.2.1->google-auth<3,>=2.14.1->hail) (0.5.1)\n",
      "Requirement already satisfied: oauthlib>=3.0.0 in /opt/conda/lib/python3.10/site-packages (from requests-oauthlib>=0.7.0->google-auth-oauthlib<1,>=0.5.2->hail) (3.2.2)\n",
      "Requirement already satisfied: pycparser in /opt/conda/lib/python3.10/site-packages (from cffi>=1.12->cryptography>=2.5->azure-identity<2,>=1.6.0->hail) (2.21)\n"
     ]
    }
   ],
   "source": [
    "!pip install hail"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "606cd6b4-7758-43b2-85a7-6ea9977d97ce",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Install the Google Cloud Storage filesystem connector\n",
    "\n",
    "The JSON key file needs to be generated in the GCP console (https://console.cloud.google.com/iam-admin/serviceaccounts/)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c896b17b-5cd1-4b6b-a592-7d9382a12260",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "curl: /opt/conda/lib/libcurl.so.4: no version information available (required by curl)\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100 10989  100 10989    0     0   145k      0 --:--:-- --:--:-- --:--:--  147k\n"
     ]
    }
   ],
   "source": [
    "!curl https://raw.githubusercontent.com/broadinstitute/install-gcs-connector/master/install_gcs_connector.py > ../install_gcs_connector.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "511d69c0-e285-4710-9e44-0c409d4fee1d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: install_gcs_connector.py [-h] [-a AUTH_TYPE] [-k KEY_FILE_PATH]\n",
      "                                [--gcs-requester-pays-project GCS_REQUESTER_PAYS_PROJECT]\n",
      "\n",
      "options:\n",
      "  -h, --help            show this help message and exit\n",
      "  -a AUTH_TYPE, --auth-type AUTH_TYPE\n",
      "                        How to authenticate. For Spark <3.5.0, this option\n",
      "                        must be unspecified. For Spark >=3.5.0, use --auth-\n",
      "                        type APPLICATION_DEFAULT for your laptop. Use --auth-\n",
      "                        type COMPUTE_ENGINE for GCE VMs. Use\n",
      "                        SERVICE_ACCOUNT_JSON_KEYFILE with --key-file-path for\n",
      "                        an explicit key file location. For Spark >=3.5.0, we\n",
      "                        default to APPLICATION_DEFAULT.\n",
      "  -k KEY_FILE_PATH, --key-file-path KEY_FILE_PATH\n",
      "                        Required for Spark <3.5.0. Service account key .json\n",
      "                        path. This path is just added to the spark config\n",
      "                        file. The .json file itself doesn't need to exist\n",
      "                        until the GCS connector is first used.\n",
      "  --gcs-requester-pays-project GCS_REQUESTER_PAYS_PROJECT, --gcs-requestor-pays-project GCS_REQUESTER_PAYS_PROJECT\n",
      "                        If specified, this google cloud project will be\n",
      "                        charged for access to requester pays buckets via\n",
      "                        spark/hadoop. See\n",
      "                        https://github.com/GoogleCloudDataproc/hadoop-\n",
      "                        connectors/blob/master/gcs/CONFIGURATION.md#cloud-\n",
      "                        storage-requester-pays-feature-configuration\n"
     ]
    }
   ],
   "source": [
    "!python3 ../install_gcs_connector.py -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c2811e0a-0753-4726-ad7f-4de096a20445",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-04-01 01:27:02,219 INFO     Downloading https://repo1.maven.org/maven2/com/google/cloud/bigdataoss/gcs-connector/hadoop2-2.2.21/gcs-connector-hadoop2-2.2.21-shaded.jar\n",
      "2024-04-01 01:27:02,219 INFO        to /opt/conda/lib/python3.10/site-packages/pyspark/jars/gcs-connector-hadoop2-2.2.21-shaded.jar\n",
      "2024-04-01 01:27:02,501 INFO     Updating /opt/conda/lib/python3.10/site-packages/pyspark/conf/spark-defaults.conf json.keyfile\n",
      "2024-04-01 01:27:02,501 INFO     Setting json.keyfile = ../psychic-rhythm-198903-23c401baac7b.json\n"
     ]
    }
   ],
   "source": [
    "!python3 ../install_gcs_connector.py -k ../psychic-rhythm-198903-23c401baac7b.json"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9387022-5df9-4d14-aabd-edc44aeac930",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Ensure that Java is installed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "7446cd86-3bd6-47fb-9831-60c35ae9dfee",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "openjdk version \"11.0.22\" 2024-01-16\n",
      "OpenJDK Runtime Environment (build 11.0.22+7-post-Debian-1deb11u1)\n",
      "OpenJDK 64-Bit Server VM (build 11.0.22+7-post-Debian-1deb11u1, mixed mode, sharing)\n"
     ]
    }
   ],
   "source": [
    "!java -version"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6c1236d-2591-4ea1-bee9-1daa47732980",
   "metadata": {},
   "source": [
    "## Test connection to gnomAD v4.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "90714e18-d87f-418b-bfb3-e1171bdb68fa",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "24/04/01 01:47:21 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting default log level to \"WARN\".\n",
      "To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).\n",
      "Running on Apache Spark version 3.3.4\n",
      "SparkUI available at http://loic-powersfs-v2-motebook.us-central1-a.c.psychic-rhythm-198903.internal:4040\n",
      "Welcome to\n",
      "     __  __     <>__\n",
      "    / /_/ /__  __/ /\n",
      "   / __  / _ `/ / /\n",
      "  /_/ /_/\\_,_/_/_/   version 0.2.128-eead8100a1c1\n",
      "LOGGING: writing to /home/jupyter/powerSFS/hail-20240401-0147-0.2.128-eead8100a1c1.log\n"
     ]
    }
   ],
   "source": [
    "import hail as hl;\n",
    "hl.init()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "882af1f4-9d91-4cba-9845-34fcb6e9bf4c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----------------------------------------\n",
      "Global fields:\n",
      "    'freq_meta': array<dict<str, str>> \n",
      "    'freq_index_dict': dict<str, int32> \n",
      "    'freq_meta_sample_count': array<int32> \n",
      "    'faf_meta': array<dict<str, str>> \n",
      "    'faf_index_dict': dict<str, int32> \n",
      "    'joint_freq_meta': array<dict<str, str>> \n",
      "    'joint_freq_index_dict': dict<str, int32> \n",
      "    'joint_freq_meta_sample_count': array<int32> \n",
      "    'joint_faf_meta': array<dict<str, str>> \n",
      "    'joint_faf_index_dict': dict<str, int32> \n",
      "    'age_distribution': struct {\n",
      "        bin_edges: array<float64>, \n",
      "        bin_freq: array<int32>, \n",
      "        n_smaller: int32, \n",
      "        n_larger: int32\n",
      "    } \n",
      "    'downsamplings': dict<str, array<int32>> \n",
      "    'filtering_model': struct {\n",
      "        filter_name: str, \n",
      "        score_name: str, \n",
      "        snv_cutoff: struct {\n",
      "            bin: int32, \n",
      "            min_score: float64\n",
      "        }, \n",
      "        indel_cutoff: struct {\n",
      "            bin: int32, \n",
      "            min_score: float64\n",
      "        }, \n",
      "        snv_training_variables: array<str>, \n",
      "        indel_training_variables: array<str>\n",
      "    } \n",
      "    'inbreeding_coeff_cutoff': float64 \n",
      "    'interval_qc_parameters': struct {\n",
      "        per_platform: bool, \n",
      "        all_platforms: bool, \n",
      "        high_qual_cutoffs: dict<str, array<tuple (\n",
      "            str, \n",
      "            str, \n",
      "            float64\n",
      "        )>>, \n",
      "        min_platform_size: int32\n",
      "    } \n",
      "    'tool_versions': struct {\n",
      "        cadd_version: str, \n",
      "        revel_version: str, \n",
      "        spliceai_version: str, \n",
      "        pangolin_version: array<str>, \n",
      "        phylop_version: str, \n",
      "        dbsnp_version: str, \n",
      "        sift_version: str, \n",
      "        polyphen_version: str\n",
      "    } \n",
      "    'vrs_versions': struct {\n",
      "        vrs_schema_version: str, \n",
      "        vrs_python_version: str, \n",
      "        seqrepo_version: str\n",
      "    } \n",
      "    'vep_globals': struct {\n",
      "        vep_version: str, \n",
      "        vep_help: str, \n",
      "        vep_config: str, \n",
      "        gencode_version: str, \n",
      "        mane_select_version: str\n",
      "    } \n",
      "    'frequency_README': str \n",
      "    'date': str \n",
      "    'version': str \n",
      "----------------------------------------\n",
      "Row fields:\n",
      "    'locus': locus<GRCh38> \n",
      "    'alleles': array<str> \n",
      "    'freq': array<struct {\n",
      "        AC: int32, \n",
      "        AF: float64, \n",
      "        AN: int32, \n",
      "        homozygote_count: int64\n",
      "    }> \n",
      "    'grpmax': struct {\n",
      "        gnomad: struct {\n",
      "            AC: int32, \n",
      "            AF: float64, \n",
      "            AN: int32, \n",
      "            homozygote_count: int64, \n",
      "            gen_anc: str\n",
      "        }, \n",
      "        non_ukb: struct {\n",
      "            AC: int32, \n",
      "            AF: float64, \n",
      "            AN: int32, \n",
      "            homozygote_count: int64, \n",
      "            gen_anc: str\n",
      "        }\n",
      "    } \n",
      "    'faf': array<struct {\n",
      "        faf95: float64, \n",
      "        faf99: float64\n",
      "    }> \n",
      "    'fafmax': struct {\n",
      "        gnomad: struct {\n",
      "            faf95_max: float64, \n",
      "            faf95_max_gen_anc: str, \n",
      "            faf99_max: float64, \n",
      "            faf99_max_gen_anc: str\n",
      "        }, \n",
      "        non_ukb: struct {\n",
      "            faf95_max: float64, \n",
      "            faf95_max_gen_anc: str, \n",
      "            faf99_max: float64, \n",
      "            faf99_max_gen_anc: str\n",
      "        }\n",
      "    } \n",
      "    'joint_freq': array<struct {\n",
      "        AC: int32, \n",
      "        AF: float64, \n",
      "        AN: int32, \n",
      "        homozygote_count: int32\n",
      "    }> \n",
      "    'joint_grpmax': struct {\n",
      "        AC: int32, \n",
      "        AF: float64, \n",
      "        AN: int32, \n",
      "        homozygote_count: int32, \n",
      "        gen_anc: str\n",
      "    } \n",
      "    'joint_faf': array<struct {\n",
      "        faf95: float64, \n",
      "        faf99: float64\n",
      "    }> \n",
      "    'joint_fafmax': struct {\n",
      "        faf95_max: float64, \n",
      "        faf95_max_gen_anc: str, \n",
      "        faf99_max: float64, \n",
      "        faf99_max_gen_anc: str, \n",
      "        joint_fafmax_data_type: str\n",
      "    } \n",
      "    'a_index': int32 \n",
      "    'was_split': bool \n",
      "    'rsid': set<str> \n",
      "    'filters': set<str> \n",
      "    'info': struct {\n",
      "        FS: float64, \n",
      "        MQ: float64, \n",
      "        MQRankSum: float64, \n",
      "        QUALapprox: int64, \n",
      "        QD: float64, \n",
      "        ReadPosRankSum: float64, \n",
      "        SB: array<int32>, \n",
      "        SOR: float64, \n",
      "        VarDP: int32, \n",
      "        AS_FS: float64, \n",
      "        AS_MQ: float64, \n",
      "        AS_MQRankSum: float64, \n",
      "        AS_pab_max: float64, \n",
      "        AS_QUALapprox: int64, \n",
      "        AS_QD: float64, \n",
      "        AS_ReadPosRankSum: float64, \n",
      "        AS_SB_TABLE: array<int32>, \n",
      "        AS_SOR: float64, \n",
      "        AS_VarDP: int32, \n",
      "        singleton: bool, \n",
      "        transmitted_singleton: bool, \n",
      "        sibling_singleton: bool, \n",
      "        omni: bool, \n",
      "        mills: bool, \n",
      "        monoallelic: bool, \n",
      "        only_het: bool, \n",
      "        AS_VQSLOD: float64, \n",
      "        inbreeding_coeff: float64, \n",
      "        vrs: struct {\n",
      "            VRS_Allele_IDs: array<str>, \n",
      "            VRS_Starts: array<int32>, \n",
      "            VRS_Ends: array<int32>, \n",
      "            VRS_States: array<str>\n",
      "        }\n",
      "    } \n",
      "    'vep': struct {\n",
      "        allele_string: str, \n",
      "        end: int32, \n",
      "        id: str, \n",
      "        input: str, \n",
      "        intergenic_consequences: array<struct {\n",
      "            allele_num: int32, \n",
      "            consequence_terms: array<str>, \n",
      "            impact: str, \n",
      "            variant_allele: str\n",
      "        }>, \n",
      "        most_severe_consequence: str, \n",
      "        motif_feature_consequences: array<struct {\n",
      "            allele_num: int32, \n",
      "            consequence_terms: array<str>, \n",
      "            high_inf_pos: str, \n",
      "            impact: str, \n",
      "            motif_feature_id: str, \n",
      "            motif_name: str, \n",
      "            motif_pos: int32, \n",
      "            motif_score_change: float64, \n",
      "            transcription_factors: array<str>, \n",
      "            strand: int32, \n",
      "            variant_allele: str\n",
      "        }>, \n",
      "        regulatory_feature_consequences: array<struct {\n",
      "            allele_num: int32, \n",
      "            biotype: str, \n",
      "            consequence_terms: array<str>, \n",
      "            impact: str, \n",
      "            regulatory_feature_id: str, \n",
      "            variant_allele: str\n",
      "        }>, \n",
      "        seq_region_name: str, \n",
      "        start: int32, \n",
      "        strand: int32, \n",
      "        transcript_consequences: array<struct {\n",
      "            allele_num: int32, \n",
      "            amino_acids: str, \n",
      "            appris: str, \n",
      "            biotype: str, \n",
      "            canonical: int32, \n",
      "            ccds: str, \n",
      "            cdna_start: int32, \n",
      "            cdna_end: int32, \n",
      "            cds_end: int32, \n",
      "            cds_start: int32, \n",
      "            codons: str, \n",
      "            consequence_terms: array<str>, \n",
      "            distance: int32, \n",
      "            domains: array<struct {\n",
      "                db: str, \n",
      "                name: str\n",
      "            }>, \n",
      "            exon: str, \n",
      "            flags: str, \n",
      "            gene_id: str, \n",
      "            gene_pheno: int32, \n",
      "            gene_symbol: str, \n",
      "            gene_symbol_source: str, \n",
      "            hgnc_id: str, \n",
      "            hgvsc: str, \n",
      "            hgvsp: str, \n",
      "            hgvs_offset: int32, \n",
      "            impact: str, \n",
      "            intron: str, \n",
      "            lof: str, \n",
      "            lof_flags: str, \n",
      "            lof_filter: str, \n",
      "            lof_info: str, \n",
      "            mane_select: str, \n",
      "            mane_plus_clinical: str, \n",
      "            mirna: array<str>, \n",
      "            protein_end: int32, \n",
      "            protein_start: int32, \n",
      "            protein_id: str, \n",
      "            source: str, \n",
      "            strand: int32, \n",
      "            transcript_id: str, \n",
      "            tsl: int32, \n",
      "            uniprot_isoform: array<str>, \n",
      "            variant_allele: str\n",
      "        }>, \n",
      "        variant_class: str\n",
      "    } \n",
      "    'vqsr_results': struct {\n",
      "        AS_VQSLOD: float64, \n",
      "        AS_culprit: str, \n",
      "        positive_train_site: bool, \n",
      "        negative_train_site: bool\n",
      "    } \n",
      "    'region_flags': struct {\n",
      "        non_par: bool, \n",
      "        lcr: bool, \n",
      "        segdup: bool, \n",
      "        fail_interval_qc: bool, \n",
      "        outside_ukb_capture_region: bool, \n",
      "        outside_broad_capture_region: bool\n",
      "    } \n",
      "    'allele_info': struct {\n",
      "        variant_type: str, \n",
      "        n_alt_alleles: int32, \n",
      "        has_star: bool, \n",
      "        allele_type: str, \n",
      "        was_mixed: bool\n",
      "    } \n",
      "    'histograms': struct {\n",
      "        qual_hists: struct {\n",
      "            gq_hist_all: struct {\n",
      "                bin_edges: array<float64>, \n",
      "                bin_freq: array<int64>, \n",
      "                n_smaller: int64, \n",
      "                n_larger: int64\n",
      "            }, \n",
      "            dp_hist_all: struct {\n",
      "                bin_edges: array<float64>, \n",
      "                bin_freq: array<int64>, \n",
      "                n_smaller: int64, \n",
      "                n_larger: int64\n",
      "            }, \n",
      "            gq_hist_alt: struct {\n",
      "                bin_edges: array<float64>, \n",
      "                bin_freq: array<int64>, \n",
      "                n_smaller: int64, \n",
      "                n_larger: int64\n",
      "            }, \n",
      "            dp_hist_alt: struct {\n",
      "                bin_edges: array<float64>, \n",
      "                bin_freq: array<int64>, \n",
      "                n_smaller: int64, \n",
      "                n_larger: int64\n",
      "            }, \n",
      "            ab_hist_alt: struct {\n",
      "                bin_edges: array<float64>, \n",
      "                bin_freq: array<int64>, \n",
      "                n_smaller: int64, \n",
      "                n_larger: int64\n",
      "            }\n",
      "        }, \n",
      "        raw_qual_hists: struct {\n",
      "            gq_hist_all: struct {\n",
      "                bin_edges: array<float64>, \n",
      "                bin_freq: array<int64>, \n",
      "                n_smaller: int64, \n",
      "                n_larger: int64\n",
      "            }, \n",
      "            dp_hist_all: struct {\n",
      "                bin_edges: array<float64>, \n",
      "                bin_freq: array<int64>, \n",
      "                n_smaller: int64, \n",
      "                n_larger: int64\n",
      "            }, \n",
      "            gq_hist_alt: struct {\n",
      "                bin_edges: array<float64>, \n",
      "                bin_freq: array<int64>, \n",
      "                n_smaller: int64, \n",
      "                n_larger: int64\n",
      "            }, \n",
      "            dp_hist_alt: struct {\n",
      "                bin_edges: array<float64>, \n",
      "                bin_freq: array<int64>, \n",
      "                n_smaller: int64, \n",
      "                n_larger: int64\n",
      "            }, \n",
      "            ab_hist_alt: struct {\n",
      "                bin_edges: array<float64>, \n",
      "                bin_freq: array<int64>, \n",
      "                n_smaller: int64, \n",
      "                n_larger: int64\n",
      "            }\n",
      "        }, \n",
      "        age_hists: struct {\n",
      "            age_hist_het: struct {\n",
      "                bin_edges: array<float64>, \n",
      "                bin_freq: array<int64>, \n",
      "                n_smaller: int64, \n",
      "                n_larger: int64\n",
      "            }, \n",
      "            age_hist_hom: struct {\n",
      "                bin_edges: array<float64>, \n",
      "                bin_freq: array<int64>, \n",
      "                n_smaller: int64, \n",
      "                n_larger: int64\n",
      "            }\n",
      "        }\n",
      "    } \n",
      "    'in_silico_predictors': struct {\n",
      "        cadd: struct {\n",
      "            phred: float32, \n",
      "            raw_score: float32\n",
      "        }, \n",
      "        revel_max: float64, \n",
      "        spliceai_ds_max: float32, \n",
      "        pangolin_largest_ds: float64, \n",
      "        phylop: float64, \n",
      "        sift_max: float64, \n",
      "        polyphen_max: float64\n",
      "    } \n",
      "----------------------------------------\n",
      "Key: ['locus', 'alleles']\n",
      "----------------------------------------\n"
     ]
    }
   ],
   "source": [
    "ht = hl.read_table('gs://gcp-public-data--gnomad/release/4.0/ht/exomes/gnomad.exomes.v4.0.sites.ht')\n",
    "ht.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "a88e3bbe-a9f9-4975-b316-e35a245c4c42",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "hl.stop()"
   ]
  }
 ],
 "metadata": {
  "environment": {
   "kernel": "conda-root-py",
   "name": "workbench-notebooks.m119",
   "type": "gcloud",
   "uri": "us-docker.pkg.dev/deeplearning-platform-release/gcr.io/workbench-notebooks:m119"
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel) (Local)",
   "language": "python",
   "name": "conda-root-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
